File: src/lib/commands/knowledge.ts
import type { KnowledgeStats, QueryResult, Suggestion } from '../../types/knowledge';
import { apiDelete, apiGet, apiPost } from '../api/client';

export async function queryKnowledge(
  repoPath: string,
  worktreeId: string | null,
  query: string,
  topK?: number,
): Promise<QueryResult[]> {
  try {
    return await apiPost<QueryResult[]>('/knowledge/query', {
      repoPath,
      worktreeId,
      query,
      topK: topK ?? null,
    });
  } catch (e) {
    console.error(`queryKnowledge failed: ${e}`);
    throw e;
  }
}

export async function ingestKnowledge(
  repoPath: string,
  worktreeId: string | null,
  content: string,
  entryType: string,
): Promise<number> {
  try {
    return await apiPost<number>('/knowledge/ingest', {
      repoPath,
      worktreeId,
      content,
      entryType,
    });
  } catch (e) {
    console.error(`ingestKnowledge failed: ${e}`);
    throw e;
  }
}

export async function getKnowledgeStats(repoPath: string): Promise<KnowledgeStats> {
  try {
    return await apiGet<KnowledgeStats>(
      `/knowledge/stats?repoPath=${encodeURIComponent(repoPath)}`,
    );
  } catch (e) {
    console.error(`getKnowledgeStats failed: ${e}`);
    throw e;
  }
}

export async function forgetKnowledge(entryId: number): Promise<void> {
  try {
    await apiDelete(`/knowledge/entries/${entryId}`);
  } catch (e) {
    console.error(`forgetKnowledge failed: ${e}`);
    throw e;
  }
}

export async function suggestNext(
  repoPath: string,
  context: string,
  topK?: number,
): Promise<Suggestion[]> {
  try {
    return await apiPost<Suggestion[]>('/knowledge/suggest', {
      repoPath,
      context,
      topK: topK ?? null,
    });
  } catch (e) {
    console.error(`suggestNext failed: ${e}`);
    throw e;
  }
}

File: src-tauri/src/commands/knowledge.rs
//! Tauri IPC commands for the knowledge service.

use crate::error::AppError;
#[cfg(feature = "knowledge")]
use crate::models::knowledge::{KnowledgeStats, KnowledgeType, QueryResult};
#[cfg(feature = "knowledge")]
use crate::services::knowledge::KnowledgeService;
#[cfg(feature = "knowledge")]
use std::sync::Arc;
#[cfg(feature = "knowledge")]
use tauri::State;

#[cfg(feature = "knowledge")]
#[tauri::command]
pub async fn query_knowledge(
    knowledge: State<'_, Arc<KnowledgeService>>,
    repo_path: String,
    worktree_id: Option<String>,
    query: String,
    top_k: Option<usize>,
) -> Result<Vec<QueryResult>, AppError> {
    let k = top_k.unwrap_or(5).min(100);
    knowledge
        .query(&repo_path, worktree_id.as_deref(), &query, k)
        .await
}

#[cfg(feature = "knowledge")]
#[tauri::command]
pub async fn ingest_knowledge(
    knowledge: State<'_, Arc<KnowledgeService>>,
    repo_path: String,
    worktree_id: Option<String>,
    content: String,
    entry_type: String,
) -> Result<u64, AppError> {
    let kt = match entry_type.as_str() {
        "commit" => KnowledgeType::Commit,
        "error_resolution" => KnowledgeType::ErrorResolution,
        "pattern" => KnowledgeType::Pattern,
        _ => KnowledgeType::Explicit,
    };

    knowledge
        .ingest_explicit(&repo_path, worktree_id.as_deref(), &content, kt)
        .await
}

#[cfg(feature = "knowledge")]
#[tauri::command]
pub async fn get_knowledge_stats(
    knowledge: State<'_, Arc<KnowledgeService>>,
    repo_path: String,
) -> Result<KnowledgeStats, AppError> {
    knowledge.get_stats(&repo_path).await
}

#[cfg(feature = "sona")]
#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub async fn suggest_next(
    knowledge: State<'_, Arc<KnowledgeService>>,
    repo_path: String,
    context: String,
    top_k: Option<usize>,
) -> Result<Vec<crate::models::knowledge::Suggestion>, AppError> {
    if context.trim().is_empty() {
        return Err(AppError::Knowledge("context must not be empty".to_string()));
    }
    let k = top_k.unwrap_or(5).min(20);
    knowledge.suggest_next(&repo_path, &context, k).await
}
